### PR TITLE
PROP-88 -  Use Semantic Definition Format (SDF) for Data and Interactions of Things

### LINKED ISSUE #88: Feature: Use Semantic Definition Format (SDF) for Data and Interactions of Things
### Is your feature request related to a problem? Please describe.

No

### Describe the feature you are requesting, as well as the possible use case(s) for it.

https://datatracker.ietf.org/doc/draft-ietf-asdf-sdf/

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by JeffMboya**
# What type of PR is this?

This is a feature because it adds SDF (Semantic Definition Format) support for describing proplet data and interactions.

## What does this do?

Introduces a new `pkg/sdf` package that implements the [Semantic Definition Format (SDF)](https://datatracker.ietf.org/doc/draft-ietf-asdf-sdf/) for describing proplet data and interactions as standardized Thing descriptions.

- Adds `pkg/sdf` with `Document` and affordance types (`PropertyAfford`, `ActionAfford`, `EventAfford`, `ObjectAfford`, `ThingAfford`, `DataAfford`) following the SDF spec
- Adds `PropletDocument(p proplet.Proplet) sdf.Document` which builds a full SDF Thing description for a proplet, including properties (`id`, `name`, `alive`, `task_count`, `metadata`), actions (`start_task`, `stop_task`), events (`heartbeat`, `task_result`), and data schemas (`PropletMetadata`, `TaskDispatch`, `HeartbeatPayload`, `TaskResult`)
- Wires `GetPropletSDF` through the full manager stack: service interface, implementation, endpoint, HTTP transport (`GET /proplets/{id}/sdf`), logging/metrics/tracing middleware, and mock
- Exposes `GetPropletSDF(id string) (sdf.Document, error)` on the SDK interface

## Which issue(s) does this PR fix/relate to?

- Resolves #88

## Have you included tests for your changes?

Yes. `pkg/sdf/proplet_test.go` contains table-driven tests covering:
- Correct `Info` fields (title contains proplet ID, non-empty version)
- All expected `sdfProperty`, `sdfAction`, `sdfEvent`, and `sdfData` keys on the `Proplet` thing
- `TaskResult.state` enum completeness (`Completed`, `Failed`, `Skipped`, `Interrupted`)
- `Minimum` pointer aliasing safety across `DataAfford` fields

## Did you document any new/modified features?

The new `GET /proplets/{id}/sdf` endpoint and `GetPropletSDF` SDK method follow the existing patterns in the codebase. The SDK interface entry includes an inline usage example consistent with all other SDK methods.

### Notes

The SDF document is returned as a flat JSON object (fields of `sdf.Document` directly at the top level), consistent with how other response types embed their domain structs. The shape matches what the SDK `GetPropletSDF` unmarshals.

### CODE REVIEW

File: pkg/sdf/sdf.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package sdf
```
Comment (rodneyosodo) [inline pkg/sdf/sdf.go:1]: **`minPtr()` placement**: This helper is only called from `proplet.go`. Consider moving it there to keep `sdf.go` a pure type definition file.

File: pkg/sdf/sdf.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+type EventAfford struct {
+	Description   string      `json:"description,omitempty"`
+	SdfOutputData *DataAfford `json:"sdfOutputData,omitempty"`
+}
+
+type ObjectAfford struct {
+	Description string                    `json:"description,omitempty"`
+	SdfProperty map[string]PropertyAfford `json:"sdfProperty,omitempty"`
```
Comment (rodneyosodo) [inline pkg/sdf/sdf.go]: **Unused `ObjectAfford` fields**: `ThingAfford.SdfObject` and `Document.SdfObject` are defined but never populated. Per project convention, remove premature scaffolding — add these when a use case exists.

File: pkg/sdf/proplet.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+							SdfRef: "#/sdfThing/Proplet/sdfData/TaskDispatch",
+						},
+					},
+					"stop_task": {
+						Description: "Stop a running task on this proplet",
+						SdfInputData: &DataAfford{
+							Description: "Identifier of the task to stop",
+							Type:        "string",
+						},
+					},
```
Comment (rodneyosodo) [inline pkg/sdf/proplet.go:68]: **Inconsistent schema for `stop_task`**: `start_task` references a named `sdfData` entry (`TaskDispatch`), but `stop_task` uses an anonymous inline `DataAfford`. Consider adding a named `TaskStop` schema for consistency:
```go
"stop_task": {
    Description: "Stop a running task on this proplet",
    SdfInputData: &DataAfford{
        SdfRef: "#/sdfThing/Proplet/sdfData/TaskStop",
    },
}
```

**Review by rodneyosodo — COMMENTED** [review]
Rebase with main

File: pkg/sdf/proplet.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+						Type:        "object",
+						Required:    []string{"id", "name"},
+						Properties: map[string]DataAfford{
+							"id":                {Type: "string"},
+							"name":              {Type: "string"},
+							"image_url":         {Type: "string"},
+							"cli_args":          {Type: "array", Items: &DataAfford{Type: "string"}},
+							"inputs":            {Type: "array", Items: &DataAfford{Type: "string"}},
+							"env":               {Type: "object", AdditionalProperties: &DataAfford{Type: "string"}},
+							"daemon":            {Type: "boolean"},
```
Comment (rodneyosodo) [inline pkg/sdf/proplet.go:113]: **Open schema for `results`**: No `Type` is set, making this an unconstrained open schema. This is valid SDF but may surprise consumers. If the output is always an array or object, consider constraining it, or at minimum expand the description to set expectations.

File: pkg/sdf/proplet_test.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
+package sdf_test
```
Comment (rodneyosodo) [inline pkg/sdf/proplet_test.go:1]: **Missing transport-level tests**: `pkg/sdf` unit tests are solid, but there are no HTTP-level tests for `GET /proplets/{id}/sdf` (e.g. 404 on unknown ID, 200 with correct JSON shape). Other endpoints in `manager/api/` have transport tests — this endpoint should too.

File: pkg/sdf/sdf.go (hunk 1 above)
Comment (JeffMboya) [reply pkg/sdf/sdf.go:1]: Moved `minPtr()` from `sdf.go` into `proplet.go`. `sdf.go` is now a pure type definition file.

File: pkg/sdf/sdf.go (hunk 2 above)
Comment (JeffMboya) [reply pkg/sdf/sdf.go]: Removed `SdfObject` from both `Document` and `ThingAfford`. No use case exists yet; scaffolding dropped.

File: pkg/sdf/proplet.go (hunk 3 above)
Comment (JeffMboya) [reply pkg/sdf/proplet.go:68]: Added named `TaskStop` sdfData entry. `stop_task` now references it via `sdfRef`, matching `start_task`'s pattern.

File: pkg/sdf/proplet.go (hunk 4 above)
Comment (JeffMboya) [reply pkg/sdf/proplet.go:113]: Set `Type: "object"` and added a description clarifying keys and value types depend on the wasm module.

File: pkg/sdf/proplet_test.go (hunk 5 above)
Comment (JeffMboya) [reply pkg/sdf/proplet_test.go:1]: Added `manager/api/transport_test.go` with table-driven tests covering 200 OK, 404 unknown ID, and 404 invalid ID.

**Review by rodneyosodo — COMMENTED** [review]
A few things to fix before merging — one test bug, one dead-code struct, and a few minor nits.

File: manager/api/transport_test.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+		{
+			desc:       "get SDF for existing proplet",
+			propletID:  validID,
+			svcDoc:     validDoc,
+			svcErr:     nil,
+			wantStatus: http.StatusOK,
+			wantTitle:  "test",
+		},
+		{
+			desc:       "get SDF for unknown proplet returns 404",
```
Comment (rodneyosodo) [inline manager/api/transport_test.go:56]: **Bug — test description contradicts `wantStatus`.**

The description says "returns 500" but `wantStatus` is `http.StatusNotFound` (404). These need to match.

Also worth noting: `validate()` on `entityReq` only checks for an empty string — it does not reject non-UUID values. So `"not-a-valid-uuid"` passes validation, reaches the service mock (which returns `ErrNotFound`), and correctly maps to 404. The test behaviour is right but the description is misleading. Consider renaming to `"get SDF for non-existent proplet ID returns 404"` and removing the implication that validation rejects it.

File: manager/api/transport_test.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+		},
+		{
+			desc:       "get SDF with invalid proplet ID returns 500",
+			propletID:  "not-a-valid-uuid",
+			svcDoc:     sdf.Document{},
+			svcErr:     pkgerrors.ErrNotFound,
+			wantStatus: http.StatusNotFound,
+		},
+	}
+
```
Comment (rodneyosodo) [inline manager/api/transport_test.go:75]: **Missing test case: empty proplet ID.**

`validate()` returns `ErrMissingID` when `id == ""`, which should map to a 4xx (likely 400). Add a case with `propletID: ""` and assert it does not reach the service.

File: pkg/sdf/sdf.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
```
Comment (rodneyosodo) [inline pkg/sdf/sdf.go]: `ObjectAfford` is defined but never used anywhere in this PR or the existing codebase. Either remove it or add a comment explaining when it would be used. Leaving it as dead code will confuse future readers, especially because it is structurally identical to `ThingAfford`.

File: pkg/sdf/sdf.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	SdfAction   map[string]ActionAfford   `json:"sdfAction,omitempty"`
+	SdfEvent    map[string]EventAfford    `json:"sdfEvent,omitempty"`
+	SdfData     map[string]DataAfford     `json:"sdfData,omitempty"`
+}
+
+type ThingAfford struct {
+	Description string                    `json:"description,omitempty"`
+	SdfProperty map[string]PropertyAfford `json:"sdfProperty,omitempty"`
```
Comment (rodneyosodo) [inline pkg/sdf/sdf.go:54]: `ThingAfford` and `ObjectAfford` are byte-for-byte identical in structure. If `ObjectAfford` is intentional (e.g. for future nested SDF objects), a type alias (`type ObjectAfford = ThingAfford`) or a shared embedded base struct would prevent them from silently diverging.

File: pkg/sdf/proplet.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
+package sdf
+
+import "github.com/absmach/propeller/pkg/proplet"
+
+func PropletDocument(p proplet.Proplet) Document {
+	return Document{
+		Info: Info{
+			Title:   "Propeller Proplet: " + p.Name + " (" + p.ID + ")",
+			Version: "1.0",
+			License: "Apache-2.0",
```
Comment (rodneyosodo) [inline pkg/sdf/proplet.go:14]: `Version: "1.0"` is hardcoded and has no tie to the proplet's version or the API version. If the SDF schema changes in future, consumers have no way to detect the old shape from the version field alone. Consider extracting this to a package-level constant (e.g. `const schemaVersion = "1.0"`) so it is easy to find and bump.

File: pkg/sdk/proplet.go [hunk 11]
```diff
@@ -4,12 +4,30 @@ import (
...
+	"github.com/absmach/propeller/pkg/sdf"
 )
 
 const propletsEndpoint = "/proplets"
 
+type Proplet struct {
+	ID        string    `json:"id"`
+	Name      string    `json:"name"`
+	TaskCount uint64    `json:"task_count"`
+	Alive     bool      `json:"alive"`
```
Comment (rodneyosodo) [inline pkg/sdk/proplet.go]: `sdk.Proplet` silently drops fields present in `proplet.Proplet` (e.g. `Metadata`, `PassKey`, `Channels`). This is intentional as a stable SDK surface — but it should be documented (a comment on the struct) so the next person editing this file does not assume the two types are interchangeable.

File: pkg/sdk/proplet.go [hunk 12]
```diff
@@ -26,6 +44,52 @@ func (sdk *propSDK) GetPropletAliveHistory(id string, offset, limit uint64) (pro
...
+	}
+
+	var doc sdf.Document
+	if err := json.Unmarshal(body, &doc); err != nil {
+		return sdf.Document{}, err
+	}
+
+	return doc, nil
+}
+
```
Comment (rodneyosodo) [inline pkg/sdk/proplet.go:68]: Skipping `offset=0` from the query string means an explicit `offset=0` call is indistinguishable from "no offset given". This matches how other SDK methods are written today so it's consistent — just flagging that if the server default ever changes this will silently break callers passing `0` explicitly. `limit=0` has the same issue and is arguably more dangerous (server might interpret it as "unlimited").

File: manager/api/transport_test.go (hunk 6 above)
Comment (JeffMboya) [reply manager/api/transport_test.go:56]: Fixed — description now says 'non-existent proplet ID returns 404'.

File: manager/api/transport_test.go (hunk 7 above)
Comment (JeffMboya) [reply manager/api/transport_test.go:75]: Added empty-ID case; also fixed EncodeError to map ErrValidation → 400.

File: pkg/sdf/sdf.go (hunk 8 above)
Comment (JeffMboya) [reply pkg/sdf/sdf.go]: Replaced with type alias `type ObjectAfford = ThingAfford` — no dead code.

File: pkg/sdf/sdf.go (hunk 9 above)
Comment (JeffMboya) [reply pkg/sdf/sdf.go:54]: Resolved via type alias; divergence is now structurally impossible.

File: pkg/sdf/proplet.go (hunk 10 above)
Comment (JeffMboya) [reply pkg/sdf/proplet.go:14]: Extracted to package-level `const schemaVersion = "1.0"`.

File: pkg/sdk/proplet.go (hunk 11 above)
Comment (JeffMboya) [reply pkg/sdk/proplet.go]: Intentional minimal SDK surface; noted, no change per no-comments policy.

File: pkg/sdk/proplet.go (hunk 12 above)
Comment (JeffMboya) [reply pkg/sdk/proplet.go:68]: Acknowledged — consistent with all other SDK methods today.

File: manager/api/transport_test.go [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+			svcDoc:     sdf.Document{},
+			svcErr:     pkgerrors.ErrNotFound,
+			wantStatus: http.StatusNotFound,
+		},
+		{
+			desc:       "get SDF for non-existent proplet ID returns 404",
+			propletID:  "not-a-valid-uuid",
+			svcDoc:     sdf.Document{},
+			svcErr:     pkgerrors.ErrNotFound,
+			wantStatus: http.StatusNotFound,
```
Comment (rodneyosodo) [inline manager/api/transport_test.go]: The expected status here is likely **400**, not 404. `decodeEntityReq` passes the path param through `entityReq.validate()`, which rejects non-UUID values with an `ErrValidation`-wrapped error — hitting the new `case errors.Is(err, apiutil.ErrValidation)` branch added in `pkg/api/api.go`. The mock for `GetPropletSDF` would never be reached in this case. Consider changing `wantStatus` to `http.StatusBadRequest` and removing the `svcErr` field from this test case.

File: pkg/sdf/sdf.go [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+	SdfAction   map[string]ActionAfford   `json:"sdfAction,omitempty"`
+	SdfEvent    map[string]EventAfford    `json:"sdfEvent,omitempty"`
+	SdfData     map[string]DataAfford     `json:"sdfData,omitempty"`
+}
+
+type ObjectAfford = ThingAfford
```
Comment (rodneyosodo) [inline pkg/sdf/sdf.go]: `ObjectAfford` is defined but never used anywhere in this package or the rest of the codebase. Remove it, or add an `sdfObject` field to `Document` and `ThingAfford` (and use it) if you actually need it.

File: pkg/sdk/proplet.go [hunk 15]
```diff
@@ -4,12 +4,30 @@ import (
...
+	"strings"
+	"time"
 
 	"github.com/absmach/propeller/pkg/proplet"
+	"github.com/absmach/propeller/pkg/sdf"
 )
 
 const propletsEndpoint = "/proplets"
 
+type Proplet struct {
```
Comment (rodneyosodo) [inline pkg/sdk/proplet.go:20]: `sdk.Proplet` and `sdk.PropletPage` duplicate `proplet.Proplet` / `proplet.PropletPage` from `pkg/proplet`. Other SDK methods (e.g. `GetPropletAliveHistory`) reuse domain types directly. If the intent is to expose a stable, separate public shape, add a comment explaining why the duplication is intentional and which fields deliberately differ — otherwise reuse the domain type to avoid silent drift.

File: pkg/sdk/proplet.go [hunk 16]
```diff
@@ -26,6 +44,52 @@ func (sdk *propSDK) GetPropletAliveHistory(id string, offset, limit uint64) (pro
...
+
+	var doc sdf.Document
+	if err := json.Unmarshal(body, &doc); err != nil {
+		return sdf.Document{}, err
+	}
+
+	return doc, nil
+}
+
+func (sdk *propSDK) ListProplets(offset, limit uint64, status string) (PropletPage, error) {
```
Comment (rodneyosodo) [inline pkg/sdk/proplet.go:69]: `ListProplets` is unrelated to SDF and has no test coverage in this PR. Either move it to a separate PR or add a `TestListProplets` transport test case before merging.

File: pkg/sdf/proplet.go [hunk 17]
```diff
@@ -0,0 +1,158 @@
+package sdf
+
+import "github.com/absmach/propeller/pkg/proplet"
+
+const schemaVersion = "1.0"
```
Comment (rodneyosodo) [inline pkg/sdf/proplet.go:7]: `schemaVersion = "1.0"` is a magic constant with no link to the proplet's actual version or the SDF spec revision. Document what this version represents (proplet schema version? SDF spec version?), or source it from a build/version constant so it doesn't silently go stale.

File: pkg/sdf/proplet_test.go [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+			for _, key := range tc.wantSdfDataKeys {
+				_, ok := thing.SdfData[key]
+				assert.True(t, ok, "expected sdfData[%q] to exist", key)
+			}
+		})
+	}
+}
+
+func TestPropletDocumentTaskResultStateEnum(t *testing.T) {
```
Comment (rodneyosodo) [inline pkg/sdf/proplet_test.go:82]: `TestPropletDocumentTaskResultStateEnum` and `TestPropletDocumentMinimumPointerNotAliased` (line 124) each wrap a single hard-coded case in a table-driven loop. The table adds ceremony without enabling variation — simplify to a plain `t.Run` or inline assertions unless you expect additional cases.

File: manager/api/transport_test.go (hunk 13 above)
Comment (JeffMboya) [reply manager/api/transport_test.go]: Fixed. Added UUID validation to entityReq.validate(); non-UUID IDs now return 400 before the service is reached.

File: pkg/sdf/sdf.go (hunk 14 above)
Comment (JeffMboya) [reply pkg/sdf/sdf.go]: Removed. ObjectAfford type alias deleted from sdf.go.

File: pkg/sdk/proplet.go (hunk 15 above)
Comment (JeffMboya) [reply pkg/sdk/proplet.go:20]: Added comments to both types explaining why the SDK surface intentionally differs from the domain types.

File: pkg/sdk/proplet.go (hunk 16 above)
Comment (JeffMboya) [reply pkg/sdk/proplet.go:69]: Added TestListProplets transport test covering 200 with results, 200 empty page, and 400 on invalid offset.

File: pkg/sdf/proplet.go (hunk 17 above)
Comment (JeffMboya) [reply pkg/sdf/proplet.go:7]: Added a comment documenting that schemaVersion tracks the proplet SDF schema definition, not the SDF spec, and should be bumped on breaking changes.

File: pkg/sdf/proplet_test.go (hunk 18 above)
Comment (JeffMboya) [reply pkg/sdf/proplet_test.go:82]: Simplified both to inline assertions, removing the single-case table wrapper.

**Review by rodneyosodo — APPROVED** [review]
